// @antv/g6/lib/plugins/menu/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _tslib = require("tslib");

var _modifyCss = _interopRequireDefault(require("@antv/dom-util/lib/modify-css"));

var _createDom = _interopRequireDefault(require("@antv/dom-util/lib/create-dom"));

var _isString = _interopRequireDefault(require("@antv/util/lib/is-string"));

var _insertCss = _interopRequireDefault(require("insert-css"));

var _base = _interopRequireDefault(require("../base"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _insertCss.default)("\n  .g6-component-contextmenu {\n    border: 1px solid #e2e2e2;\n    border-radius: 4px;\n    font-size: 12px;\n    color: #545454;\n    background-color: rgba(255, 255, 255, 0.9);\n    padding: 10px 8px;\n    box-shadow: rgb(174, 174, 174) 0px 0px 10px;\n  }\n  .g6-contextmenu-ul {\n    padding: 0;\n    margin: 0;\n    list-style: none;\n  }\n\n");

var Menu =
/** @class */
function (_super) {
  (0, _tslib.__extends)(Menu, _super);

  function Menu(cfg) {
    return _super.call(this, cfg) || this;
  }

  Menu.prototype.getDefaultCfgs = function () {
    return {
      handleMenuClick: undefined,
      // 指定菜单内容，function(e) {...}
      getContent: function getContent(graph) {
        return "\n          <ul class='g6-contextmenu-ul'>\n            <li>\u83DC\u5355\u98791</li>\n            <li>\u83DC\u5355\u98792</li>\n          </ul>\n        ";
      },
      // 菜单隐藏事件
      onHide: function onHide() {
        return true;
      }
    };
  }; // class-methods-use-this


  Menu.prototype.getEvents = function () {
    return {
      contextmenu: 'onMenuShow'
    };
  };

  Menu.prototype.init = function () {
    var className = this.get('className');
    var menu = (0, _createDom.default)("<div class=" + (className || 'g6-component-contextmenu') + "></div>");
    (0, _modifyCss.default)(menu, {
      position: 'absolute',
      visibility: 'hidden'
    });
    var container = this.get('container');

    if (!container) {
      container = this.get('graph').get('container');
    }

    container.appendChild(menu);
    this.set('menu', menu);
  };

  Menu.prototype.onMenuShow = function (e) {
    e.preventDefault();
    e.stopPropagation();

    if (!e.item) {
      return;
    }

    var self = this;
    var menuDom = this.get('menu');
    var getContent = this.get('getContent');
    var menu = getContent(e);

    if ((0, _isString.default)(menu)) {
      menuDom.innerHTML = menu;
    } else {
      menuDom.innerHTML = menu.outerHTML;
    }

    var handleMenuClick = this.get('handleMenuClick');

    if (handleMenuClick) {
      menuDom.addEventListener('click', function (evt) {
        handleMenuClick(evt.target, e.item);
      });
    }

    var graph = this.get('graph');
    var width = graph.get('width');
    var height = graph.get('height');
    var bbox = menuDom.getBoundingClientRect();
    var x = e.canvasX; //e.item.getModel().x || e.x;

    var y = e.canvasY; //e.item.getModel().y || e.y;
    // 若菜单超出画布范围，反向

    if (x + bbox.width > width) {
      x = width - bbox.width;
    }

    if (y + bbox.height > height) {
      y = height - bbox.height;
    } // const point = graph.getClientByPoint(x, y)
    // e.canvasX = point.x;
    // e.canvasY = point.y;


    (0, _modifyCss.default)(menuDom, {
      top: y + "px",
      left: x + "px",
      visibility: 'visible'
    });

    var handler = function handler(evt) {
      self.onMenuHide();
    }; // 如果在页面中其他任意地方进行click, 隐去菜单


    document.body.addEventListener('click', handler);
    this.set('handler', handler);
  };

  Menu.prototype.onMenuHide = function () {
    var menuDom = this.get('menu');

    if (menuDom) {
      (0, _modifyCss.default)(menuDom, {
        visibility: 'hidden'
      });
    } // 隐藏菜单后需要移除事件监听


    document.body.removeEventListener('click', this.get('handler'));
    var handleMenuClick = this.get('handleMenuClick');

    if (handleMenuClick) {
      menuDom.removeEventListener('click', handleMenuClick);
    }
  };

  Menu.prototype.destroy = function () {
    var menu = this.get('menu');
    var handler = this.get('handler');
    var handleMenuClick = this.get('handleMenuClick');

    if (handleMenuClick) {
      menu.removeEventListener('click', handleMenuClick);
    }

    if (menu) {
      var container = this.get('container');

      if (!container) {
        container = this.get('graph').get('container');
      }

      container.removeChild(menu);
    }

    if (handler) {
      document.body.removeEventListener('click', handler);
    }
  };

  return Menu;
}(_base.default);

var _default = Menu;
exports.default = _default;
